### ISSUE TITLE
Clarification on Gradient Accumulation with AutoModel

### ISSUE THREAD

**Issue by Kurokabe** [NVIDIA/NeMo#12964]
Hi team,

I've been experimenting with the new NeMo `AutoModel` to train a HuggingFace model, and I’m trying to understand the correct way to configure gradient accumulation in this setup.

Previously, using the standard NeMo pipeline, I relied on a `LightningDataModule` with a `MegatronDataSampler`. Gradient accumulation was automatically handled by the `MegatronStrategy`, depending on the `micro_batch_size`, `global_batch_size`, and the number of devices. For example:

```python
class MyDataModule(pl.LightningDataModule):
    def __init__(
        self,
        micro_batch_size: int = 4,
        global_batch_size: int = 8,
    ):
        self.data_sampler = MegatronDataSampler(
            micro_batch_size=micro_batch_size,
            global_batch_size=global_batch_size,
        )
```

However, with `AutoModel`, I'm unclear on how these batch size parameters are meant to work. Specifically:

- The `nemo.lightning.pytorch.strategies.FSDP2Strategy` accepts a `data_sampler` argument, but it doesn't seem to influence gradient accumulation behavior.
- Similarly, `nemo.collections.llm.gpt.data.hf_dataset.HFDatasetDataModule` exposes `micro_batch_size` and `global_batch_size` parameters, but changing `global_batch_size` appears to have no effect.

Could you clarify the correct way to configure `micro_batch_size` and `global_batch_size` when using `AutoModel`? Should we be handling gradient accumulation manually, or is there a mechanism similar to what exists with the `MegatronStrategy`?

Thanks!

**Comment by akoumpa** [comment]
Hi @Kurokabe , 

For the gradient accumulation please use the [accumulate_grad_batches](https://github.com/NVIDIA/NeMo/blob/d0fc65838ba2bc6f2506da742d5af427acfd0747/examples/llm/peft/automodel.py#L223) argument in `nl.Trainer`'s constructor, and for the batch size please use the [micro_batch_size](https://github.com/NVIDIA/NeMo/blob/d0fc65838ba2bc6f2506da742d5af427acfd0747/examples/llm/peft/automodel.py#L58) argument only.

Initially, we added support for Megatron datasets in AutoModel path and it made sense to have both options (micro_batch_size and global_batch_size) available, however, for technical reasons this was later removed.

We are planning (probably in 25.06) to have a more streamlined api for the dataset and gradient accumulation options.

Please let me know if that helps or you still face this issue.

I apologize for the inconvenience and I'll update the docs to reflect this divergence in the api.

Thank you.

**Comment by Kurokabe** [comment]
Thank you @akoumpa for your answer, this works well with the `accumulate_grad_batches` parameter.
Is there also any plan to add more `MegatronStrategy` functionalities with the AutoModel, like async checkpoint saving?

**Comment by akoumpa** [comment]
@Kurokabe yes, while the current checkpoint functionality is limited, we are planning to support async checkpointing in the future.
